update the webauthn crate

stefan0xC opened the pull request:
I have been working on updating the `webauthn_rs` crate (as basis to implement passkey support).

The feature `danger-allow-state-serialisation` is used to serialize the state into the database, which should be fine according to: https://docs.rs/webauthn-rs/0.4.8/webauthn_rs/index.html#allow-serialising-registration-and-authentication-state

I've also decided to remove the u2f migration because it would have required the use of the more low level, protocol interactions provided by the [webauthn_core_rs](https://docs.rs/webauthn-rs-core/latest/webauthn_rs_core/) crate, so I guess this could be considered a breaking change? (If this is deemed necessary I can revert the removal, I mainly did it because upgrading the crate was tedious enough.)

I could not actually test the changes because I don't have a security key myself. And therefore I also don't know if it addresses the issues raised in https://github.com/dani-garcia/vaultwarden/issues/4196 (but I think it should be easier with the use of the Safe API?). So someone else definitely needs to test it and/or take over this PR.

BlackDex commented:
The whole migration part is why both me and @dani-garcia probably did not finished this yet too.

I do think we need to create a migration, else we should release a 2.0 or something like that.

Without the migration it will either lock out people, or make there account less secure if you would ignore the old version. 

Also, regarding testing, can't you use Bitwarden passkeys?
Else you could try (if using Linux at least) https://github.com/danstiner/rust-u2f

stefan0xC commented:
> I do think we need to create a migration, else we should release a 2.0 or something like that.

Well, I don't think I've already chipped into the discussion [about breaking changes and semantic versioning](https://github.com/dani-garcia/vaultwarden/issues/4023) but to me doing a `2.0` release (and communicating that you'd have to re-register any FIDO2 webauthn compatible security key) would probably be my preferred option (at the moment) if it means that we don't have to do this work at all (especially if the intention is to make our implementation more secure and RFC compliant).

> Without the migration it will either lock out people, or make there account less secure if you would ignore the old version.

You mean the U2F migration? Or is there a migration needed for compatibility between the way it was done and the "correct" way (according to RFC or at least the linked PR). Because if the latter I'd probably have to look a bit more into the details of the WebAuthn RFC some more before I can propose a solution myself. I mean, maybe we should store the webauthn registrations in a separate table to better ensure correctness.

> Also, regarding testing, can't you use Bitwarden passkeys?

I'm not sure that passkeys are handled the same as security keys (at least there are 4 separate functions in the `webauthn_rs` crate for handling passkey registration and authentication). So I don't think so.

> Else you could try (if using Linux at least) https://github.com/danstiner/rust-u2f

I probably could. Thanks for the suggestion.

stefan0xC commented:
I've reverted the removal of the U2F migration but I have not checked if I missed anything.
I also did not have time to test it either so that's why I've changed the status of the PR to WIP for now.

stefan0xC commented:
Just tested it with Chromium's built-in [WebAuthn DevTools](https://developer.chrome.com/docs/devtools/webauthn) and it's not working. I'll try to find some time to fix this but it will probably take me a bit.

BlackDex commented:
@stefan0xC, could you update this to have it rebased? That way i would also be able to test it using the Bitwarden Native Android app without any issues.

BlackDex commented:
Well, it is at least easier to check and validate now.

BlackDex commented:
Did a quick check, but it breaks. The `CredentialsV3` and `CredentialsV5` are not compatible.
There already is a migration functions which runs to migrate very old keys to the `CredentialsV3` format.

Maybe we can use that same function or an other function to do the same for `CredentialsV5`?
It currently at least reports that `user_verified` is missing during the json from_str().

I think you can test this by using The Bitwarden Browser Extension to create a Passkey which can be used as a WebAuthn login using a the current main branch, and then switch to this PR and test it out.

stefan0xC commented:
I don't really have the time to work on this, so going to close it for now.
